api/handlers.go

```diff
@@ -865,27 +941,33 @@ func (api *API) userRecoverGET(w http.ResponseWriter, req *http.Request, _ httpr
 // They need to provide a valid password-reset token.
 // The user doesn't need to be logged in.
 func (api *API) userRecoverPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
-	if err := req.ParseForm(); err != nil {
-		api.WriteError(w, err, http.StatusBadRequest)
+	defer api.closeBody(req)
+	// Parse the request's body.
+	var payload accountRecoveryDTO
+	b, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to parse request body"), http.StatusBadRequest)
+		return
+	}
+	err = json.Unmarshal(b, &payload)
+	if err != nil {
+		api.WriteError(w, errors.AddContext(err, "failed to parse request body"), http.StatusBadRequest)
 		return
 	}
-	recoveryToken := req.Form.Get("token")
-	password := req.Form.Get("password")
-	confirmPassword := req.Form.Get("confirmPassword")
-	if recoveryToken == "" || password == "" || confirmPassword == "" {
+	if payload.Password == "" || payload.ConfirmPassword == "" || payload.Token == "" {
 		api.WriteError(w, errors.New("missing required parameter"), http.StatusBadRequest)
 		return
 	}
-	if password != confirmPassword {
+	if payload.Password != payload.ConfirmPassword {
 		api.WriteError(w, errors.New("passwords don't match"), http.StatusBadRequest)
 		return
 	}
-	u, err := api.staticDB.UserByRecoveryToken(req.Context(), recoveryToken)
+	u, err := api.staticDB.UserByRecoveryToken(req.Context(), payload.Token)
 	if err != nil {
 		api.WriteError(w, errors.New("no such user"), http.StatusBadRequest)
 		return
 	}
-	passHash, err := hash.Generate(password)
+	passHash, err := hash.Generate(payload.Password)
```

Comment:
shouldn't we assert the `payload.Password` is not an empty string here? We do it on update as well right

Reply:
We do that on line 950 where we have
```go
	if payload.Password == "" || payload.ConfirmPassword == "" || payload.Token == "" {
		api.WriteError(w, errors.New("missing required parameter"), http.StatusBadRequest)
		return
	}
```